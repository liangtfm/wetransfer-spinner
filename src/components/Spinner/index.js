import React from "react";
import styled, { keyframes } from "styled-components";

const Circle = styled.circle`
  transition: stroke-dashoffset 0.35s;
  transform: rotate(-90deg);
  transform-origin: 50% 50%;
`;

const ProgressNumber = styled.span`
  font-size: 32px;
  position: absolute;
`;

const RotateKeyframes = keyframes`
  from {
    transform: rotate(0deg);
  }

  to {
    transform: rotate(360deg);
  }
`;

const Rotate = styled.div`
  display: flex;
  animation: ${RotateKeyframes} 1s cubic-bezier(0.17, 0.67, 0.83, 0.67) infinite;
  padding: 2rem 1rem;
  font-size: 1.2rem;
`;

const SpinnerContainer = styled.div`
  align-items: center;
  border-radius: 6px;
  display: flex;
  justify-content: center;
  padding: 0;
`;

function Spinner(props) {
  const color = props.color || "#409fff";
  const progress = props.progress || 0;
  const radius = props.radius || 100;
  const stroke = props.stroke || 8;
  const normalizedRadius = radius - stroke * 2;
  const circumference = normalizedRadius * 2 * Math.PI;
  const strokeDashoffset = circumference - (progress / 100) * circumference;

  return (
    <SpinnerContainer role="spinner">
      <Rotate>
        <svg height={radius * 2} width={radius * 2}>
          <Circle
            stroke={color}
            fill="transparent"
            strokeWidth={stroke}
            strokeDasharray={circumference + " " + circumference}
            style={{ strokeDashoffset }}
            stroke-width={stroke}
            r={normalizedRadius}
            cx={radius}
            cy={radius}
          />
        </svg>
      </Rotate>
      <ProgressNumber>{progress}%</ProgressNumber>
    </SpinnerContainer>
  );
}

export default Spinner;
