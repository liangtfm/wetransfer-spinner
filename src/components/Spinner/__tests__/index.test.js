import React from "react";
import { render, screen } from "@testing-library/react";

import Spinner from "..";

describe("Spinner", () => {
  it("renders correctly", () => {
    const { asFragment } = render(<Spinner />);

    expect(asFragment(<Spinner />)).toMatchSnapshot();
    expect(screen.getByText(/0%/i)).toBeInTheDocument();
  });
});
