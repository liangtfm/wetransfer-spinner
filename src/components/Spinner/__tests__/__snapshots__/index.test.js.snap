// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spinner renders correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-hKgILt guuNrP"
    role="spinner"
  >
    <div
      class="sc-dlfnbm jUdWyn"
    >
      <svg
        height="200"
        width="200"
      >
        <circle
          class="sc-bdfBwQ grbpYE"
          cx="100"
          cy="100"
          fill="transparent"
          r="84"
          stroke="#409fff"
          stroke-dasharray="527.7875658030853 527.7875658030853"
          stroke-width="8"
          style="stroke-dashoffset: 527.7875658030853;"
        />
      </svg>
    </div>
    <span
      class="sc-gsTCUz jPwBUC"
    >
      0%
    </span>
  </div>
</DocumentFragment>
`;
